File: src/RelaxedIK/Utils_Julia/transformations.jl
using Rotations, StaticArrays, LinearAlgebra

q = Quat(1.0,0.0,0.0,0.0)
q2 = Quat(0.0,1.0,0.0,0.0)

function quaternion_log(quaternion)
    out_vec = MVector(quaternion.x, quaternion.y, quaternion.z)
    if abs(quaternion.w) < 1.0
        a = 1
        a = acos(quaternion.w)
        sina = sin(a)
        if abs(sina >= 0.005)
            c = a/sina
            out_vec[1] *= c
            out_vec[2] *= c
            out_vec[3] *= c
        end
    end
    return out_vec
end

function quaternion_exp(vec3)
    q = MVector(1.0, vec3[1], vec3[2], vec3[3])
    a = norm(q)
    sina = sin(a)
    if abs(sina) >= 0.005
        c = sina/a
        q[2]*=c
        q[3]*=c
        q[4]*=c
    end
    q[1] = cos(a)
    return Quat(q[1], q[2], q[3], q[4])
end

function quaternion_disp(q, qPrime)
    return quaternion_log(inv(q)*qPrime)
end

function quaternion_dispQ(q, qPrime)
    return inv(q)*qPrime
end

# using BenchmarkTools

# @btime quaternion_dispQ(q, q2)

File: src/RelaxedIK_Rust/src/bin/lib/utils_rust/transformations.rs
use nalgebra::{Vector3, UnitQuaternion, Quaternion, Vector4};

pub fn quaternion_log(q: UnitQuaternion<f64>) -> Vector3<f64> {
    let mut out_vec: Vector3<f64> = Vector3::new(q.i, q.j, q.k);
    if q.w.abs() < 1.0 {
        let a = q.w.acos();
        let sina = a.sin();
        if sina.abs() >= 0.005 {
            let c = a / sina;
            out_vec *= c;
        }
    }
    out_vec
}

pub fn quaternion_exp(v: Vector3<f64>) -> UnitQuaternion<f64> {
    let mut qv: Vector4<f64> = Vector4::new(1.0, v[0], v[1], v[2]);
    let a = qv.norm();
    let sina = a.sin();
    if sina.abs() >= 0.005 {
        let c = sina/a;
        qv *= c;
    }
    UnitQuaternion::from_quaternion(Quaternion::new(a.cos(), qv[1], qv[2], qv[3]))
}

pub fn quaternion_disp(q: UnitQuaternion<f64>, q_prime: UnitQuaternion<f64>) -> Vector3<f64> {
    quaternion_log( q.inverse()*q_prime )
}

pub fn quaternion_dispQ(q: UnitQuaternion<f64>, q_prime: UnitQuaternion<f64>) -> UnitQuaternion<f64> {
    q.inverse()*q_prime
}

pub fn angle_between(q: UnitQuaternion<f64>, q_prime: UnitQuaternion<f64>) -> f64 {
    quaternion_disp(q, q_prime).norm() * 2.0
}
